File: 3_duplicate_number/main.py
# -*-  coding: utf-8 -*-


def duplicate(nums):
    if len(nums) <= 0:
        return False, 0

    for i in range(len(nums)):
        if nums[i] < 0 or nums[i] >= len(nums):
            return False, 0
        while nums[i] != i:
            if nums[i] == nums[nums[i]]:
                return True, nums[i]
            nums[nums[0]], nums[0] = nums[0], nums[nums[0]]
    return False


if __name__ == '__main__':
    nums = [2, 3, 1, 0, 2, 5, 3]
    duplicated, dup_num = duplicate(nums)
    if not duplicated:
        print('no duplicated number')
    else:
        print('duplicated number is {0}'.format(dup_num))

File: 3_duplicate_number/go/main.go
package main

import (
	"fmt"
)

func Duplicate(nums []int) (bool, int) {
	size := len(nums)
	if size <= 0 {
		return false, 0
	}

	for i := range nums {
		if nums[i] < 0 || nums[i] >= size {
			return false, 0
		}
		for nums[i] != i {
			if nums[i] == nums[nums[i]] {
				return true, nums[i]
			}
			nums[nums[i]], nums[i] = nums[i], nums[nums[i]]
		}
	}

	return false, 0
}

func main() {
	nums := []int{2, 3, 1, 0, 2, 5, 3}
	duplicated, dupNum := Duplicate(nums)
	if !duplicated {
		fmt.Println("no duplicated number")
	} else {
		fmt.Printf("duplicated number is %d\n", dupNum)
	}
}
